chainerx_cc/chainerx/cuda/memory_pool.cc

```diff
@@ -107,39 +217,57 @@ void* MemoryPool::Malloc(size_t bytesize) {
                 throw OutOfMemoryError{bytesize};
             }
         }
+        chunk = std::make_unique<Chunk>(ptr, 0, allocation_size);
     }
 
+    CHAINERX_ASSERT(chunk != nullptr);
+    void* ptr = chunk->ptr();
     {
         std::lock_guard<std::mutex> lock{in_use_mutex_};
-        in_use_.emplace(ptr, index);
+        in_use_.emplace(ptr, std::move(chunk));
     }
-
     return ptr;
 }
 
 void MemoryPool::Free(void* ptr) {
     if (ptr == nullptr) {
         return;
     }
-    size_t index{};
+    std::unique_ptr<Chunk> chunk{nullptr};
 
     {
         std::lock_guard<std::mutex> lock{in_use_mutex_};
         auto it = in_use_.find(ptr);
         if (it == in_use_.end()) {
             throw ChainerxError{"Cannot free out-of-pool memory"};
         }
-        index = it->second;
+        chunk = std::move(it->second);
         in_use_.erase(it);
     }
 
+    CHAINERX_ASSERT(chunk != nullptr);
     {
         std::lock_guard<std::mutex> lock{free_bins_mutex_};
-        if (free_bins_.size() <= index) {
-            free_bins_.resize(index + 1);
+
+        // If there are free chunks before this chunk, merges it with them.
+        if (chunk->next() != nullptr) {
+            std::unique_ptr<Chunk> chunk_next = RemoveChunkFromFreeList(chunk->next());
+            if (chunk_next != nullptr) {
+                chunk->MergeWithNext();
+            }
+        }
+
+        // If there are free chunks after this chunk, merges it with them.
+        if (chunk->prev() != nullptr) {
+            std::unique_ptr<Chunk> chunk_prev = RemoveChunkFromFreeList(chunk->prev());
+            if (chunk_prev != nullptr) {
+                chunk_prev->MergeWithNext();
+                chunk = std::move(chunk_prev);
+            }
         }
-        std::vector<void*>& free_list = free_bins_[index];
-        free_list.emplace_back(ptr);
+
+        // Pushes this chunk into the free list.
```

Comment:
NITPICK: This comment says same thing with the function name. So, I prefer to remove the redundant information, but this is just my preference.